feat: Import onyx state

TMisiukiewicz opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR adds "Import onyx state" feature, making it possible to load other user account state in an offline mode in order to track and fix hardly reproducible bugs/track performance improvements.
### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/47786
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Go to Settings -> Troubleshoot and Export Onyx State
2. Log out from the account
3. Log into another account
4. Go to Settings -> Troubleshoot
5. Press "Import Onyx State"
6. Select the file exported from another account
7. Wait for the onyx state to load
8. Verify the red button saying the state is imported is displayed
9. Verify app is in offline mode
10. Navigate to a few screens and verify app is usable
11. Press the red button
12. Verify the account is back online and is using the Onyx state from the original account
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
n/a
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
13. Upload an image via copy paste
14. Verify a modal appears displaying a preview of that image
--->
1. Go to Settings -> Troubleshoot and Export Onyx State
2. Log out from the account
3. Log into another account
4. Go to Settings -> Troubleshoot
5. Press "Import Onyx State"
6. Select the file exported from another account
7. Wait for the onyx state to load
8. Verify the red button saying the state is imported is displayed
9. Verify app is in offline mode
10. Navigate to a few screens and verify app is usable
11. Press the red button
12. Verify the account is back online and is using the Onyx state from the original account
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/c7f0040c-c016-436c-8338-fbd9db454584


</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/819694d2-ffa9-4359-93ac-8e1afa49b53e


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/02a3367e-8851-47e7-8ac2-c06758f3f870


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/2757e193-1f07-476d-b7df-7192c026b0e8


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/0a5bde60-cc83-43eb-aecf-e8fbb3f89a02


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/f304b4d2-c734-4b4b-a7ff-0eedacb31c3c


<!-- add screenshots or videos here -->

</details>

adhorodyski committed:
fix: improve perf of isActionOfType by limiting calls to the includes

adhorodyski committed:
feat(wip): draft the import-onyx-state perf debug functionality

adhorodyski committed:
chore: add comments describing the whole workflow in steps

kubabutkiewicz committed:
fix: not iterable items from onyx in persisted requests, added web support, changed the way for selecting onyx keys, instead for including used omitting

kubabutkiewicz committed:
Merge branch 'main' of github.com:callstack-internal/Expensify-App into feat/import-onyx-state

kubabutkiewicz committed:
fix: make the android component more performant to not crash the app when importing state

kubabutkiewicz committed:
Merge branch 'main' of github.com:callstack-internal/Expensify-App into feat/import-onyx-state

adhorodyski committed:
Merge branch 'Expensify:main' into feat/import-onyx-state

adhorodyski committed:
Merge branch 'Expensify:main' into feat/import-onyx-state

kubabutkiewicz committed:
feat: add returning to original state

kubabutkiewicz committed:
Merge branch 'feat/import-onyx-state' of github.com:callstack-internal/Expensify-App into feat/import-onyx-state

kubabutkiewicz committed:
fix: clear onyx state before applying imported one

kubabutkiewicz committed:
fix: exporting onyx state on native

kubabutkiewicz committed:
fix: refreshing page

TMisiukiewicz committed:
Merge branch 'main' into feat/import-onyx-state

TMisiukiewicz committed:
align import button on web to be a menu item

TMisiukiewicz committed:
align import onyx state button on mobile as menu item

TMisiukiewicz committed:
adjust transformNumericKeysToArray function usage

TMisiukiewicz committed:
adjust web implementation

TMisiukiewicz committed:
adjust mobile import, cleanup mobile implementation

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
display only file upload on native

TMisiukiewicz committed:
force offline when importing on mobile

TMisiukiewicz committed:
reduce diff across platforms

TMisiukiewicz committed:
add onyx key to watch if state is imported

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
make sure loader is hidden

TMisiukiewicz committed:
code cleanup

TMisiukiewicz committed:
add imported state button for resetting onyx state

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
fix crash when reportID is empty after clearing state

TMisiukiewicz committed:
fix lint

kacper-mikolajczak commented on src/components/ImportOnyxState/index.native.tsx:
```diff
@@ -0,0 +1,125 @@
+                return fileContent;
+            });
+        })
+        .catch((error) => {
+            throw error;
+        });
```
Is the catch clause needed if we re-throw the error?

kacper-mikolajczak commented on src/components/ImportOnyxState/index.native.tsx:
```diff
@@ -0,0 +1,125 @@
+    const entries = Object.entries(state);
+    const chunks = chunkArray(entries, CHUNK_SIZE);
+
+    let promise = Promise.resolve();
+    chunks.forEach((chunk) => {
+        const partialOnyxState = Object.fromEntries(chunk) as Partial<OnyxValues>;
+        promise = promise.then(() => Onyx.multiSet(partialOnyxState));
+    });
+
+    return promise;
```
Correct me if I am wrong but it looks like we are overriding the promise from previous chunk with a promise from next chunk. Is this desired behaviour?

What happens with promises generated from chunks before the very last one?

TMisiukiewicz replied:
it creates a new promise that waits for the previous promise to resolve before applying the next chunk, creating a chain of promises. This ensures that the `Onyx.multiSet` operation for each chunk will execute in sequence, waiting for the previous operation to complete

kacper-mikolajczak replied:
Oh right, that's nice!

Thanks for explanation ❤️

kacper-mikolajczak commented on src/components/ImportOnyxState/utils.ts:
```diff
@@ -0,0 +1,57 @@
+}
+
+function transformNumericKeysToArray(data: UnknownRecord): UnknownRecord | unknown[] {
+    const dataCopy = data;
```
It looks like it's missing copying the `data` object.

kacper-mikolajczak reviewed (changes requested):
Nice piece of functionality right there, GJ @TMisiukiewicz 💪 

Haven't gone through all the files yet but I hope included remarks be helpful.

TMisiukiewicz committed:
code review updates

DylanDylann commented:
@TMisiukiewicz Are there any restrictions on the file size? I'm getting an error while uploading an Onyx file

<img width="1512" alt="Screenshot 2024-09-19 at 15 15 19" src="https://github.com/user-attachments/assets/a2f674fa-4cba-4cf8-84d6-5690dae6247c">

DylanDylann commented:
Using this file to reproduce 
[onyx-state (6).txt](https://github.com/user-attachments/files/17055897/onyx-state.6.txt)

TMisiukiewicz committed:
fix error when importing the file

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz commented:
@DylanDylann thanks for finding it, could you try again please? Should be fixed now. There are no restrictions about the file size

TMisiukiewicz committed:
migrate from withOnyx to useOnyx

TMisiukiewicz committed:
fix typecheck

TMisiukiewicz committed:
Merge branch 'main' into feat/import-onyx-state

DylanDylann commented:
@TMisiukiewicz I still can't use features when importing Onyx file. Chats don't display any data 


https://github.com/user-attachments/assets/f516e953-956d-4e66-ae7f-3a9d22f0dc23

TMisiukiewicz commented:
@DylanDylann just to confirm, did you visit this particular chat before exporting state, so all the messages are exported in the file?

DylanDylann commented:
@TMisiukiewicz 

1. Many comments display `fragment.text`
2. The composer is hide
3. The avatar on the profile page isn't display

https://github.com/user-attachments/assets/17069f6c-5c75-4685-b651-0825ba42d38a

TMisiukiewicz commented:
Thanks, taking a look

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
fix lint

TMisiukiewicz commented:
I just exported pretty big state (19 mb) and succesfully loaded it into the app, couldn't reproduce any of these issues, mind sharing your file so i can compare the data?

https://github.com/user-attachments/assets/b92f43b6-7e1a-4355-8ce0-9336a9aa4b40

DylanDylann commented:
[onyx-state (13).txt](https://github.com/user-attachments/files/17113594/onyx-state.13.txt)

@TMisiukiewicz Let's try this file

TMisiukiewicz committed:
fix displaying messages and composer

TMisiukiewicz commented:
@DylanDylann comments displaying `fragment.text` and hidden composer should be fixed now, I'm still looking into issue with displaying the avatar

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
update App actions

TMisiukiewicz committed:
fix onyx set error

TMisiukiewicz committed:
don't restore sequential queue when going back from imported state

TMisiukiewicz commented:
The avatar and user data are based on the `session` and `personalDetailsList` keys. We intentionally don’t clear the session to allow returning to the original user account with the correct authentication key. This design displays the original account’s avatar and data if the accountID from the `session` exists in `personalDetailsList`. However, if the `accountID` is not found (as is the case with your account), skeletons are displayed due to incomplete data. Since this does not impact the app’s functionality in import mode, I suggest keeping it as is rather than adding some import-specific logic to handle all accounts.

Additionally, I’ve fixed one console.error and adjusted the clearing behavior to align with the recent merge in https://github.com/Expensify/App/pull/49588. The sequential queue is cleared to prevent any API calls when reverting from the imported state, because we don't want to make any API calls.

DylanDylann commented on src/pages/home/report/ReportActionCompose/ReportActionCompose.tsx:
```diff
@@ -564,22 +557,18 @@ function ReportActionCompose({
                         {hasExceededMaxCommentLength && <ExceededCommentLength />}
                     </View>
                 </OfflineWithFeedback>
+                {isUsingImportedState && !isSmallScreenWidth && (
```
```suggestion
                {!isSmallScreenWidth && (
```

DylanDylann replied:
isUsingImportedState condition already implemented in ImportedStateIndicator Component, right?

DylanDylann commented on src/components/ImportOnyxState/index.tsx:
```diff
@@ -0,0 +1,69 @@
+                        setIsLoading(false);
+                    });
+            });
+        });
```
Let's add a catch block to handle exceptional (same to native file)

DylanDylann replied:
In case users upload wrong files, we can display a modal to announce that the upload failed:

Similar to : https://github.com/Expensify/App/blob/bc106fab216fbf60cd55004a94223cacf5afa8d9/src/pages/workspace/categories/WorkspaceCategoriesPage.tsx#L432

TMisiukiewicz replied:
that's a good idea, I have added the modal 👍

DylanDylann commented:
BUG: On mWeb, the Camera option displays when importing Onyx file (We can fix this if it is simple or we also can skip it because this is minor and this feature is only in dev env)

https://github.com/user-attachments/assets/f156cbea-3582-43b5-8a6a-217965fceded

DylanDylann commented:
When the user reset to the original account, I think we should navigate them to the HOME Screen. @TMisiukiewicz What do you think?

DylanDylann commented:
@TMisiukiewicz Everything else looks good to me.

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
properly get reportNameValuePairs

TMisiukiewicz commented on src/pages/home/report/ReportActionItem.tsx:
```diff
@@ -178,7 +179,7 @@ function ReportActionItem({
     const theme = useTheme();
     const styles = useThemeStyles();
-    const [reportNameValuePairs] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_NAME_VALUE_PAIRS}${report?.reportID ?? -1}`);
+    const [reportNameValuePairs] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_NAME_VALUE_PAIRS}${!isEmpty(report?.reportID) ? report?.reportID : -1}`);
```
Oh yeah `|| -1` should be enough, I tried too much to make it work without disabling an eslint rule 😄 This change is required because when clearing the state with an open report, at some point we are receiving an empty report object where `reportID` is an empty string. It was causing the entire app crashing, since it was trying to access `reportNameValuePairs_` key. I added a comment explaining it together with disabled rule

TMisiukiewicz committed:
remove unnecessary rendering condition

TMisiukiewicz committed:
create base component and display modal when file is invalid

TMisiukiewicz committed:
navigate to home on clear, do not import preferred theme

TMisiukiewicz commented:
@DylanDylann 

> Not a serious bug, but I see an interesting problem. The background color changes when we import a new Onyx file although ONYXKEYS.PREFERRED_THEME is preserved when clearing Onyx data (We can fix this if it is simple or we also can skip it because this is minor and this feature is only in dev env)

I'd say to not import preferred theme key at all, since I don't think it can impact the feature purpose in any way. I already added it to the keys omitted during import, let me know what you think

> When the user resets to the original account, I think we should navigate them to the HOME Screen and maybe display the loading indicator while waiting for data to be loaded.

I have added navigating to home screen when pressing red info button, but left the "Clear cache and restart" in Troubleshoot behavior unchanged to stay consistent with current implementation when clearing without imported state

> BUG: On mWeb, the Camera option displays when importing Onyx file (We can fix this if it is simple or we also can skip it because this is minor and this feature is only in dev env)

I think the video you attached is not relevant to this 🤔 Could you please verify? However, if you mean the camera option from attachment picker, I couldn't reproduce it on iOS and Android mWeb, I am moved directly to file selection there

The rest of the comments are resolved, thanks  for your review!

DylanDylann commented:
@TMisiukiewicz 
- ESLint check / ESLint check failed

DylanDylann commented on src/pages/home/report/ReportActionItem.tsx:
```diff
@@ -178,7 +179,8 @@ function ReportActionItem({
     const styles = useThemeStyles();
-    const [reportNameValuePairs] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_NAME_VALUE_PAIRS}${report?.reportID ?? -1}`);
+    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing -- This is needed to prevent the app from crashing when the app is using imported state.
+    const [reportNameValuePairs] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_NAME_VALUE_PAIRS}${report?.reportID || -1}`);
```
```suggestion
    const [reportNameValuePairs] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_NAME_VALUE_PAIRS}${report?.reportID || '-1'}`);
```

DylanDylann commented on src/pages/home/report/ReportActionItem.tsx:
```diff
@@ -1,3 +1,4 @@
+import isEmpty from 'lodash/isEmpty';
```
```suggestion
```

DylanDylann commented:
@TMisiukiewicz 

> The rest of the comments are resolved, thanks for your review!

I don't see that 😄

TMisiukiewicz commented:
OMG I forgot to push 🤦

TMisiukiewicz committed:
default reportID as string

TMisiukiewicz committed:
fix linter

DylanDylann commented:
> BUG: On mWeb, the Camera option displays when importing Onyx file (We can fix this if it is simple or we also can skip it because this is minor and this feature is only in dev env)


https://github.com/user-attachments/assets/1b8b21ad-b6ff-441b-bfe6-0f0b12d12b17


@TMisiukiewicz It may be due to devices. Anyways, It's so minor, let's skip it

DylanDylann commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/27ff4f3a-95ff-4546-aeef-9781ac28de80



</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/56058598-1d6e-4d20-86cb-dc292d4b1429


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/3638371e-7460-4583-a7d5-503860496fb8


</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/f8147f7a-6f3a-469b-8848-7dc13d135a1f


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/1886714e-d46e-4ba3-a826-2a78e3f479d5



</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/fde95598-b9fb-441f-bd58-1f14ae80ee88


</details>

DylanDylann reviewed (approved):
LGTM

DylanDylann commented:
@TMisiukiewicz Great job 💯

rlinoz commented on src/components/ImportOnyxState/utils.ts:
```diff
@@ -0,0 +1,53 @@
+import ONYXKEYS from '@src/ONYXKEYS';
+
+// List of Onyx keys from the .txt file we want to keep for the local override
+const keysToOmit = [ONYXKEYS.ACTIVE_CLIENTS, ONYXKEYS.BETAS, ONYXKEYS.FREQUENTLY_USED_EMOJIS, ONYXKEYS.NETWORK, ONYXKEYS.CREDENTIALS, ONYXKEYS.SESSION, ONYXKEYS.PREFERRED_THEME];
```
I think it would make sense to export betas as well, since having or not having a beta may be the source of a bug, was this discussed somewhere?

TMisiukiewicz replied:
can betas differ across different accounts? If yes, let's definitely import them.

I was checking it on a few accounts of mine and each one of them had the same array of betas, that's why my initial thought was that it's safe to omit it

rlinoz replied:
Ah right, it can definitely be different sometimes

rlinoz reviewed (commented):
This is a really nice feature, thanks for working on it!

I have a couple of questions:

1. Should I be able to see the reports? Everytime I open then in the imported state I get a `Uh-oh, something went wrong!`
2. Should we turn off the import mode if the user turns off offline mode as well?

TMisiukiewicz commented:
@rlinoz 
> Should I be able to see the reports? Everytime I open then in the imported state I get a Uh-oh, something went wrong!

yes, you should be able to see those that are stored in the exported state, the app should not crash so it looks like some kind of an edge case we haven't stumbled upon earlier 🤔 could you share the file you are importing so I can take a look?

> Should we turn off the import mode if the user turns off offline mode as well?

yeah that's a good idea, I'll implement it 👍

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

rlinoz commented:
Yeah, I like the idea of disabling it, let's do it! We can improve later in case it gets confusing.

Here is the file:
[onyx-state(2).txt](https://github.com/user-attachments/files/17209702/onyx-state.2.txt)

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/import-onyx-state

TMisiukiewicz committed:
disable force offline when in import mode

TMisiukiewicz committed:
remove betas from omitted keys

TMisiukiewicz committed:
fix data transformation in export

TMisiukiewicz commented:
@rlinoz thank you! Could you please export onyx state one more time and try to import it? Seems like there was some unexpected data transformation happening while exporting state with masked fragile data. I also disabled the "Force offline" switch as mentioned above and removed betas from preserved keys

DylanDylann commented:
@rlinoz Could you please help to export your Onyx file again? I retest with some accounts and It works well

rlinoz commented:
Before I test on the build above let me explain what is happening in a dev account that I have
<img width="625" alt="image" src="https://github.com/user-attachments/assets/1ec444ff-7d96-49ca-bb42-9d381086354f">

Something is null in my keys but [this line that isn't changed in this PR](https://github.com/Expensify/App/blob/3e305c360254b849962917a2be19a07c99424145/src/libs/ExportOnyxState/common.ts#L43) evaluates null as object and pass it on causing the crash.

Now, I don't know if this is some weird state that I am in because it is a dev account or something else, I will try on the build above with staging accounts now.

TMisiukiewicz committed:
handle null value when masking data

TMisiukiewicz commented:
@rlinoz I couldn't reproduce that but i added an additional check for null value when exporting state, mind checking again?

rlinoz reviewed (approved):
Working great, thanks!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.
